internal/charmstore/stats.go

```diff
@@ -492,38 +492,52 @@ var LegacyDownloadCountsEnabled = true
 
 // ArchiveDownloadCounts calculates the aggregated download counts for
 // a charm or bundle.
-func (s *Store) ArchiveDownloadCounts(id *router.ResolvedURL) (thisRevision, allRevisions AggregatedCounts, err error) {
-	// TODO (frankban): remove this condition when removing the legacy counts
-	// logic.
-	if LegacyDownloadCountsEnabled {
-		return s.legacyDownloadCounts(id)
-	}
-
+func (s *Store) ArchiveDownloadCounts(id *charm.Reference) (thisRevision, allRevisions AggregatedCounts, err error) {
 	// Retrieve the aggregated stats.
-	thisRevision, err = s.aggregateStats(EntityStatsKey(&id.URL, params.StatsArchiveDownload), false)
+	thisRevision, err = s.aggregateStats(EntityStatsKey(id, params.StatsArchiveDownload), false)
 	if err != nil {
 		err = errgo.Notef(err, "cannot get aggregated count for the specific revision")
 		return
 	}
-	noRevisionId := id.URL
+	noRevisionId := *id
 	noRevisionId.Revision = -1
 	allRevisions, err = s.aggregateStats(EntityStatsKey(&noRevisionId, params.StatsArchiveDownload), true)
 	if err != nil {
 		err = errgo.Notef(err, "cannot get aggregated count for all revisions")
 		return
 	}
+	// TODO (frankban): remove this condition when removing the legacy counts
+	// logic.
+	if LegacyDownloadCountsEnabled {
+		legacyRevision, legacyAll, err := s.legacyDownloadCounts(id)
```

Comment:
perhaps it might be worth doing this concurrently with the other requests.
in fact ISTM that all three requests could be done concurrently, which
would probably speed things up by a factor of roughly three.